### PR TITLE
convert: rework ftype heuristics

### LINKED ISSUE #18182: Feature Request: convert_hf_to_gguf.py to default to the original precision
### Prerequisites

- [x] I am running the latest code. Mention the version if possible as well.
- [x] I carefully followed the [README.md](https://github.com/ggml-org/llama.cpp/blob/master/README.md).
- [x] I searched using keywords relevant to my issue to make sure that I am creating a new issue that is not already open (or closed).
- [x] I reviewed the [Discussions](https://github.com/ggml-org/llama.cpp/discussions), and have a new and useful enhancement to share.

### Feature Description

when one runs convert_hf_to_gguf.py , the resulting GGUF shpuld be in the exact same precision as the HF model being converted.

### Motivation

Currently, if a model is trained in BF16 (a very common case), convert_hf_to_gguf.py does a downcast to FP16. This can cause artifacts. In rare cases it breaks a model entirely, see #18145 . I have also seen non-obvious artifacts such as different inference results on CPU vs GPU.

I would suggest keeping the existing precision when using the GGUF conversion, and then people can quantize from the original precision as needed.

### Possible Implementation

--outtype bf16 is already supported so this seems to be a matter of flipping some defaults?

### PR DESCRIPTION
**Pull request by taronaeo**
fixes: #18182

This PR updates the heuristic detection logic for the default ftype. When `--outtype` is not specified, the heuristics will attempt to figure out the highest-fidelity 16-bit ftype based on the first tensor.

If the first tensor does not match the following, it will continue to the second and nth tensor until it finds a tensor that matches:
1. At least a 2D tensor
2. Tensor not F32 dtype

If all tensors do not match the criteria above, it will default to `f16` ftype.

Tested against the following models:
1. Granite-4.0-1B, defaulted to bf16 ftype (correct)
2. GPT-NeoX-20B, defaulted to f16 ftype (correct)

Note about alternative methods, such as relying on `config.json` dtype: some finetunes or quantisation lie about the correct dtype, so it can't be trusted. And some models like GPT-OSS do not actually contain a `dtype` key within their config.json, so the easier route was to do heuristics.

AI Declaration: AI was used when creating this PR to identify existing logic relating to heuristics and to scaffold the code.

### CODE REVIEW

**Comment by taronaeo** [comment]
I was wondering if we should warn the user about using `--outtype f16` when the model is trained using `bfloat16` since this issue came up with Granite 4.0. It would be good to stop these occurrences of `???????` or having faulty models distributed online.

**Comment by CISC** [comment]
Is it really necessary to check all the tensors? Wasn't the issue just that it defaulted to `f16` instead of `auto` (and that `auto` only really checks for `f16`)?

The reason I'm saying is because this duplicates tensor loading logic that really needs to be refactored, see https://github.com/ggml-org/llama.cpp/pull/18043#pullrequestreview-3587715561

Apart from MXFP4/FP8 models I can't remember seeing any safetensors with mixed datatypes...

**Comment by pwilkin** [comment]
@CISC once I finish my refactoring of `convert_hf_to_gguf.py` it won't really make that much of a difference :) I don't think it's that complicated, it's just like 10 lines of code.

**Review by pwilkin — CHANGES_REQUESTED** [review]
The heuristics should work a bit differently - you should only take into account (a) at least 2D tensors (b) non-F32 tensors.

**Comment by taronaeo** [comment]
> Is it really necessary to check all the tensors?

~I was a bit unsure whether the first tensor gave enough information about the `dtype` to determine the correct type. But looking at how many tensors there are for models bigger than 1B, I guess reverting back to the first tensor still makes sense.~

Edit: Ignore what I said. Will revert to using first tensor :)

**Comment by taronaeo** [comment]
> The heuristics should work a bit differently - you should only take into account (a) at least 2D tensors (b) non-F32 tensors.

In retrospect, I think looping through all the tensors might not have been a good idea, especially if the model is large. In this case I'll revert back to using the first tensor and I guess, check to see if it's `tensor.dim >= 2` and `tensor.dtype != torch.float32`.

If the conditions don't meet, we'll just jump to the next tensor and check again. Let me know your thoughts about this :)

**Comment by taronaeo** [comment]
Updated this PR and retained the original `--outtype auto` logic whereby it will only choose the highest-fidelity 16-bit ftype. Heuristics now check the first tensor for either f16 or bf16 dtype and if it doesn't match, it will continue until it gets a match.

If there are no matches, it defaults to f16 ftype. Updated the PR description also, PTAL again.

File: convert_hf_to_gguf.py [hunk 1]
```diff
@@ -141,16 +141,24 @@ def __init__(self, dir_model: Path, ftype: gguf.LlamaFileType, fname_out: Path,
...
+        # Apply heuristics to figure out typical tensor encoding based on first tensor's dtype
+        # NOTE: can't use field "torch_dtype" in config.json, because some finetunes lie.
         if self.ftype == gguf.LlamaFileType.GUESSED:
-            # NOTE: can't use field "torch_dtype" in config.json, because some finetunes lie.
-            _, first_tensor = next(self.get_tensors())
-            if first_tensor.dtype == torch.float16:
-                logger.info(f"choosing --outtype f16 from first tensor type ({first_tensor.dtype})")
-                self.ftype = gguf.LlamaFileType.MOSTLY_F16
+            for _, tensor in self.get_tensors():
+                if tensor.dim() < 2 and tensor.dtype == torch.float32:
```
Comment (CISC) [inline convert_hf_to_gguf.py]:
```suggestion
                if tensor.dim() < 2:
```

**Review by CISC — APPROVED** [review]
Might want to enumerate and have some threshold before giving up, but optional.

**Comment by taronaeo** [comment]
> Might want to enumerate and have some threshold before giving up, but optional.

Hmm ideally it should detect the correct 16-bit ftype within the first few tensors. But I don't know what a good threshold is to set it such that it doesn't fail prematurely, and doesn't take too long to fall back to f16 ftype.

I'll leave it as-is until we get a report about it taking way too long, then I can see what kind of model causes it :)

**Comment by taronaeo** [comment]
@pwilkin Re-requesting your review before we can merge :)

**Review by pwilkin — APPROVED** [review]
Aight, first F16 tensor as heuristics should be fine, can't think of an example where that would be bad (it can be F32 but that'll get skipped over, so it should cover all cases).
